### PR TITLE
Refactor database package

### PR DESCRIPTION
**Pull request by eddieajau**
This is mostly Omar's database refactoring work (and Rob's getting it back up to date). It was getting untenable to keep it up to date with all the changes so I just did an export and then imported the files over an up to date checkout. This probably warrants a fair bit of review as it touches a lot of files.

### CODE REVIEW

File: libraries/joomla/database/driver.php [hunk 1]
```diff
@@ -0,0 +1,1664 @@
...
+			}
+
+			// If the class still doesn't exist we have nothing left to do but look at the next type.  We did our best.
+			if (!class_exists($class))
+			{
+				continue;
+			}
+
+			// Sweet!  Our class exists, so now we just need to know if it passes it's test method.
+			if (call_user_func_array(array($class, 'test'), array()))
```
Comment (oc666) [inline libraries/joomla/database/driver.php:220]: We can use $class->test if we are using at least PHP 5.3

File: libraries/joomla/database/driver.php [hunk 2]
```diff
@@ -0,0 +1,1664 @@
...
+	/**
+	 * Splits a string of multiple queries into an array of individual queries.
+	 *
+	 * @param   string  $sql  Input SQL string with which to split into individual queries.
+	 *
+	 * @return  array  The queries from the input string separated into an array.
+	 *
+	 * @since   11.1
+	 */
+	public static function splitSql($sql)
```
Comment (oc666) [inline libraries/joomla/database/driver.php:295]: Someone can explain what this function used for? example would be great.

Comment (mbabker) [inline libraries/joomla/database/driver.php:295]: During install and update, this function is used to separate multiple queries in an extension's .sql file into single queries to be executed one at a time.

Comment (vietvh) [inline libraries/joomla/database/driver.php:295]: We are having _splitQueries() (in the installation folder) which do the same task.

Comment (mbabker) [inline libraries/joomla/database/driver.php:295]: The installation app itself is in need of some work (Rouven and Nikolai have both been doing some cleanup and experimenting).  This is just one of those instances where the duplicate code needs removed.

**Comment by mbabker** [comment]
The JDatabase mock needs updated.  This is why the tests are failing.

**Comment by LouisLandry** [comment]
This still isn't tracking file history appropriately.  I'm gonna take a crack at it and see if we can get that done a little better.

**Comment by chdemko** [comment]
see #803

**Comment by eddieajau** [comment]
Closing this pull. See #906

**Pull request closed without merging**